[Streams 🌊] Add processors validation and simulation gate

tonyghiani opened the pull request:
## 📓 Summary

Closes https://github.com/elastic/streams-program/issues/66

This work adds changes to prevent invalid processors from being submitted.
The main rule is that a simulation is performed before any add/edit submission to guarantee that the processor config is valid.

This work also updates the simulation API to detect whether there is a non-additive change in any simulated document.

@patpscal error reporting UI for add/edit is different since the simulator is not visible for edit, I used a callout but we can easily update this once there is a final design in place.

### Form validation + simulation

https://github.com/user-attachments/assets/f7fc351b-6efc-4500-8490-b7f1c85139bf

### Non-additive processors

https://github.com/user-attachments/assets/47b5b739-c2cf-4a74-93a8-6ef43521c7d4



<!--ONMERGE {"backportTargets":["8.x"]} ONMERGE-->

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!

tonyghiani commented:
/ci

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/266898)
* Commit: 1a0af3fd0db7f2e389662673892d6c8f0221bfa3

### Failed CI Steps
* [FTR Configs #13](https://buildkite.com/elastic/kibana-pull-request/builds/266898#019464c7-400a-4516-84d4-c0ff6ee11749)
* [FTR Configs #28](https://buildkite.com/elastic/kibana-pull-request/builds/266898#019464c7-4018-419a-b2ba-fcca4e0b2811)
* [FTR Configs #28](https://buildkite.com/elastic/kibana-pull-request/builds/266898#019464ee-b349-45d5-b323-f7b280f5c17e)
* [FTR Configs #37](https://buildkite.com/elastic/kibana-pull-request/builds/266898#019464c7-401f-42b2-a3b9-8d8d595aba30)
* [FTR Configs #47](https://buildkite.com/elastic/kibana-pull-request/builds/266898#019464c7-4027-47ed-91db-5cfa2790c074)
* [FTR Configs #90](https://buildkite.com/elastic/kibana-pull-request/builds/266898#019464c7-404a-4a2c-9915-d2f5e8b93432)
* [FTR Configs #103](https://buildkite.com/elastic/kibana-pull-request/builds/266898#019464c7-4055-46a8-be26-f71233e24014)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/266898#019464c7-4018-419a-b2ba-fcca4e0b2811) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/266898/jobs/019464c7-4018-419a-b2ba-fcca4e0b2811/artifacts/019464ee-8cbe-4e02-9ff2-b1c7137e5f8b) FTR Configs #<span></span>28 / GenAI - Knowledge Base Entries APIs @<span></span>ess Security AI Assistant - Indices with `semantic_text` fields should return all existing indices with `semantic_text` fields
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/266898#019464ee-b349-45d5-b323-f7b280f5c17e) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/266898/jobs/019464ee-b349-45d5-b323-f7b280f5c17e/artifacts/019464f6-c89e-4f8e-950d-41a8841110c8) FTR Configs #<span></span>28 / GenAI - Knowledge Base Entries APIs @<span></span>ess Security AI Assistant - Indices with `semantic_text` fields should return all existing indices with `semantic_text` fields

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/266898","number":266898,"commit":"1a0af3fd0db7f2e389662673892d6c8f0221bfa3"}],"number":266898}
buildkite-pr-comment-->

flash1293 reviewed (commented):
Looks like this is breaking updating an existing processor:

<img width="577" alt="Screenshot 2025-01-14 at 16 35 01" src="https://github.com/user-attachments/assets/38223e6d-fa33-4dae-b068-b39b34bbda9c" />

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/266975)
* Commit: fb35d27a5120b66c285620f6a12c7f62b8127e1a

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/5ad3ef7b0a76968a31b0824bd08f9132c374c841) | [after](https://github.com/elastic/kibana/commit/fb35d27a5120b66c285620f6a12c7f62b8127e1a) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 265 | 268 | +3 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/5ad3ef7b0a76968a31b0824bd08f9132c374c841) | [after](https://github.com/elastic/kibana/commit/fb35d27a5120b66c285620f6a12c7f62b8127e1a) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 267.8KB | 270.6KB | +2.8KB |

<details><summary>Unknown metric groups</summary>

#### ESLint disabled in files
| id | [before](https://github.com/elastic/kibana/commit/5ad3ef7b0a76968a31b0824bd08f9132c374c841) | [after](https://github.com/elastic/kibana/commit/fb35d27a5120b66c285620f6a12c7f62b8127e1a) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 2 | 3 | +1 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/5ad3ef7b0a76968a31b0824bd08f9132c374c841) | [after](https://github.com/elastic/kibana/commit/fb35d27a5120b66c285620f6a12c7f62b8127e1a) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 5 | 6 | +1 |


</details>


### History
* :broken_heart: [Build #266898](https://buildkite.com/elastic/kibana-pull-request/builds/266898) failed 1a0af3fd0db7f2e389662673892d6c8f0221bfa3

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/266975","number":266975,"commit":"fb35d27a5120b66c285620f6a12c7f62b8127e1a"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/266898","number":266898,"commit":"1a0af3fd0db7f2e389662673892d6c8f0221bfa3"}],"number":266975}
buildkite-pr-comment-->

tonyghiani commented:
> Here I'm picking double as the field type which will break during ingest because `"this"` can't be parsed, but the UI doesn't tell me. It could do this by including the mapping update in the simulate call. 

Make sense, the current simulation is just simulating the pipeline, not a data ingestion, updating the simulate API to try an ingest simulation should also handle this. I'll open a follow up PR for this specifically.

> Also we should show a more aggressive warning when the user disabled "Ignore failures for this processor" and there are unmatched sample documents - while unmatched docs are benign as long as failures are ignored, they are a pretty big problem once they lead to failure for ingesting the doc.

Sounds good, @patpscal can we add to design a variant that handles this case? Not sure what feedback we could provide here, something like a bold warning icon+text next to the disabled flag seems fine to me.

> Looks like this is breaking updating an existing processor

Mmm I see what's happening, to validate that the updated processor configuration is correct, I'm running a new simulation on submit retrieving the data sample, but it's possible the user won't have data ingested for the time range used.

Actually, this might be limiting also when running the validation for creating the processor. I'll take a look at how we can handle this, but I'm not aware of any JS API available in Kibana to validate a processor config without simulating ingestion/pipeline or updating a real pipeline, I believe that syntax-check lives in Elasticsearch only.

flash1293 commented:
> Actually, this might be limiting also when running the validation for creating the processor. I'll take a look at how we can handle this, but I'm not aware of any JS API available in Kibana to validate a processor config without simulating ingestion/pipeline or updating a real pipeline, I believe that syntax-check lives in Elasticsearch only.

I think we can't safely do this, because existing already processed documents might fail with the new configuration even if it would work with raw documents coming from the source. I think we just shouldn't do the validation for the editing case, only for the new processor case. Annoying, but all coming back to the same root problem that we need original documents to test edits to processing, which we can't get.

tonyghiani commented:
@flash1293 I addressed your suggestions, this is ready for another look.

@patpscal I made this to provide a warning caption once the `ignore_failure` option is disable, let me know if looks good or we can implement something else 👌 

https://github.com/user-attachments/assets/69ea8e92-2dda-4789-8e28-bea33dcc0998

gsoldevila reviewed (approved):
Not sure why we own `@kbn/object-utils`, TBD

Kerry350 commented on x-pack/solutions/observability/plugins/streams_app/public/components/stream_detail_enrichment/flyout/ignore_toggles.tsx:
```diff
@@ -0,0 +1,56 @@
+      name="ignore_missing"
+      label={i18n.translate(
+        'xpack.streams.streamDetailView.managementTab.enrichment.processorFlyout.ignoreMissingLabel',
+        { defaultMessage: 'Ignore missing' }
```
Nit: With the way labels get used in accessibility contexts should we go “Ignore missing fields” here too as it’s less ambiguous?

tonyghiani replied:
I used the same copies+help text from the ingest pipeline management, here the thing is that the "Ignore missing" refers to ignore "documents" with the missing field... so maybe "Ignore missing fields" is also not accurate, wdyt? I guess the help caption here is used specifically for this precisation.

Kerry350 replied:
Gotcha, thanks for the explanation. Let's just leave it as it's consistent with the pipeline management 👌

Kerry350 reviewed (commented):
Code LGTM 👌 I went through with the latest changes, but I’ll defer to Joe on the approval with the edge cases.

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/267240)
* Commit: 98d763b254fedcc551de93bfcf9c422ebac9293b

### Failed CI Steps
* [Jest Integration Tests #5](https://buildkite.com/elastic/kibana-pull-request/builds/267240#01946921-4b97-4227-9497-8e7ca5b7f37c)
* [FTR Configs #65](https://buildkite.com/elastic/kibana-pull-request/builds/267240#01946921-4bcf-41d8-8930-2c292530ddc0)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/267240#01946921-4bcf-41d8-8930-2c292530ddc0) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/267240/jobs/01946921-4bcf-41d8-8930-2c292530ddc0/artifacts/01946935-d5a2-4600-81be-3fdfec212bfb) FTR Configs #<span></span>65 / alerting api integration security and spaces enabled Alerts - Group 1 alerts backfill rule runs backfill api key invalidation should wait to invalidate API key until backfill for rule is complete
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/267240#01946921-4b97-4227-9497-8e7ca5b7f37c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/267240/jobs/01946921-4b97-4227-9497-8e7ca5b7f37c/artifacts/01946943-ddb6-4e96-9852-88969e4c2e39) Jest Integration Tests #<span></span>5 / StorageIndexAdapter when indexing into a clean Elasticsearch instance after rolling over the index manually and indexing the same document deletes the document from the rolled over index
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/267240#01946921-4b97-4227-9497-8e7ca5b7f37c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/267240/jobs/01946921-4b97-4227-9497-8e7ca5b7f37c/artifacts/01946943-ddb6-41e5-ae56-5dd4049bc422) Jest Integration Tests #<span></span>5 / StorageIndexAdapter when indexing into a clean Elasticsearch instance after rolling over the index manually and indexing the same document puts the document in the new write index

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/5ad3ef7b0a76968a31b0824bd08f9132c374c841) | [after](https://github.com/elastic/kibana/commit/98d763b254fedcc551de93bfcf9c422ebac9293b) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 265 | 269 | +4 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/5ad3ef7b0a76968a31b0824bd08f9132c374c841) | [after](https://github.com/elastic/kibana/commit/98d763b254fedcc551de93bfcf9c422ebac9293b) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 267.8KB | 270.0KB | +2.3KB |

<details><summary>Unknown metric groups</summary>

#### ESLint disabled in files
| id | [before](https://github.com/elastic/kibana/commit/5ad3ef7b0a76968a31b0824bd08f9132c374c841) | [after](https://github.com/elastic/kibana/commit/98d763b254fedcc551de93bfcf9c422ebac9293b) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 2 | 3 | +1 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/5ad3ef7b0a76968a31b0824bd08f9132c374c841) | [after](https://github.com/elastic/kibana/commit/98d763b254fedcc551de93bfcf9c422ebac9293b) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 5 | 6 | +1 |


</details>


### History
* :green_heart: [Build #266975](https://buildkite.com/elastic/kibana-pull-request/builds/266975) succeeded fb35d27a5120b66c285620f6a12c7f62b8127e1a
* :broken_heart: [Build #266898](https://buildkite.com/elastic/kibana-pull-request/builds/266898) failed 1a0af3fd0db7f2e389662673892d6c8f0221bfa3

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/267240","number":267240,"commit":"98d763b254fedcc551de93bfcf9c422ebac9293b"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/266975","number":266975,"commit":"fb35d27a5120b66c285620f6a12c7f62b8127e1a"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/266898","number":266898,"commit":"1a0af3fd0db7f2e389662673892d6c8f0221bfa3"}],"number":267240}
buildkite-pr-comment-->

tonyghiani commented:
> Beyond the stuff we already noted down as follow ups - for the "unmatched" filter it doesn't really make sense to keep the columns for the fields that were not extracted

Addressed with [refactor(streams): hide detected fields columns for unmatched filter](https://github.com/elastic/kibana/pull/206566/commits/a49d3cf017784923db01880a2e76f20ddbaf69e2)

https://github.com/user-attachments/assets/050a30d6-6e1b-46b8-be4d-87b1da5c7126

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/267307)
* Commit: 3ffac030667d97445d716967ac211b3e98196c07

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/4304e21933e92a4f12ef0b99d687c8d3f0bb22d0) | [after](https://github.com/elastic/kibana/commit/3ffac030667d97445d716967ac211b3e98196c07) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 269 | 273 | +4 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/4304e21933e92a4f12ef0b99d687c8d3f0bb22d0) | [after](https://github.com/elastic/kibana/commit/3ffac030667d97445d716967ac211b3e98196c07) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 269.2KB | 271.6KB | +2.4KB |

<details><summary>Unknown metric groups</summary>

#### ESLint disabled in files
| id | [before](https://github.com/elastic/kibana/commit/4304e21933e92a4f12ef0b99d687c8d3f0bb22d0) | [after](https://github.com/elastic/kibana/commit/3ffac030667d97445d716967ac211b3e98196c07) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 2 | 3 | +1 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/4304e21933e92a4f12ef0b99d687c8d3f0bb22d0) | [after](https://github.com/elastic/kibana/commit/3ffac030667d97445d716967ac211b3e98196c07) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 5 | 6 | +1 |


</details>


### History
* :yellow_heart: [Build #267240](https://buildkite.com/elastic/kibana-pull-request/builds/267240) was flaky 98d763b254fedcc551de93bfcf9c422ebac9293b
* :green_heart: [Build #266975](https://buildkite.com/elastic/kibana-pull-request/builds/266975) succeeded fb35d27a5120b66c285620f6a12c7f62b8127e1a
* :broken_heart: [Build #266898](https://buildkite.com/elastic/kibana-pull-request/builds/266898) failed 1a0af3fd0db7f2e389662673892d6c8f0221bfa3

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/267307","number":267307,"commit":"3ffac030667d97445d716967ac211b3e98196c07"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/267240","number":267240,"commit":"98d763b254fedcc551de93bfcf9c422ebac9293b"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/266975","number":266975,"commit":"fb35d27a5120b66c285620f6a12c7f62b8127e1a"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/266898","number":266898,"commit":"1a0af3fd0db7f2e389662673892d6c8f0221bfa3"}],"number":267307}
buildkite-pr-comment-->

kibanamachine commented:
Starting backport for target branches: 8.x

https://github.com/elastic/kibana/actions/runs/12791225692

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/206803">](https://github.com/elastic/kibana/pull/206803)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
